snuba/web/delete_query.py

```diff
@@ -27,23 +32,70 @@ def construct_condition(body: Dict[str, Any]) -> Expression:
     return combine_and_conditions(and_conditions)
 
 
+def _get_rows_to_delete(
+    storage_key: StorageKey, select_query_to_count_rows: Query
+) -> int:
+    formatted_select_query_to_count_rows = format_query(select_query_to_count_rows)
+    select_query_results = (
+        get_storage(storage_key)
+        .get_cluster()
+        .get_reader()
+        .execute(formatted_select_query_to_count_rows)
+    )
+    return typing.cast(int, select_query_results["data"][0]["count"])
+
+
+def _enforce_max_rows(delete_query: Query) -> None:
+    """
+    The cost of a lightweight delete operation depends on the number of matching rows in the WHERE clause and the current number of data parts.
+    This operation will be most efficient when matching a small number of rows, **and on wide parts** (where the `_row_exists` column is stored
+    in its own file)
+
+    Because of the above, we want to limit the number of rows one deletes at a time. The `MaxRowsEnforcer` will query clickhouse to see how many
+      rows we plan on deleting and if it crosses the `max_rows_to_delete` set for that storage we will reject the query.
+    """
+    select_query_to_count_rows = Query(
+        selected_columns=[
+            SelectedExpression("count", FunctionCall("count", "count", ())),
+        ],
+        from_clause=delete_query.get_from_clause(),
+        condition=delete_query.get_condition(),
+    )
+    storage_key = delete_query.get_from_clause().storage_key
+    rows_to_delete = _get_rows_to_delete(
+        storage_key=storage_key, select_query_to_count_rows=select_query_to_count_rows
+    )
+    max_rows_allowed = (
+        get_storage(storage_key).get_deletion_settings().max_rows_to_delete
+    )
+    if rows_to_delete > max_rows_allowed:
+        raise TooManyDeleteRowsException(
+            f"Too many rows to delete ({rows_to_delete}), maximum allowed is {max_rows_allowed}"
+        )
+
+
 def delete_query(
     storage: WritableTableStorage, table: str, body: Dict[str, Any]
 ) -> Result:
     cluster_name = storage.get_cluster().get_clickhouse_cluster_name()
     on_cluster = literal(cluster_name) if cluster_name else None
+    from_clause = Table(
```

Comment:
I'd also put this back in the Query, we don't need `from_clause` and `condition` to be defined separately imo